FaunaDB.Client/Reflection/Encoder.cs

```diff
@@ -0,0 +1,172 @@
+﻿using System;
+using System.Runtime.CompilerServices;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using FaunaDB.Query;
+using FaunaDB.Types;
+using FaunaDB.Attributes;
+
+namespace FaunaDB.Reflection
+{
+    public static class Encoder
+    {
+        public static Expr EncodeObject(object obj) =>
+            new EncoderImpl().Encode(obj);
+    }
+
+    class EncoderImpl
+    {
+        Stack<object> stack = new Stack<object>();
+
+        public Expr Encode(object obj)
+        {
+            if (obj == null)
+                return NullV.Instance;
+
+            if (typeof(Expr).IsInstanceOfType(obj))
+                return (Expr)obj;
+
+            if (stack.Contains(obj))
+                throw new InvalidOperationException($"Self referencing loop detected for object `{obj}`");
+
+            try
+            {
+                stack.Push(obj);
+
+                return EncodeIntern(obj);
+            }
+            finally
+            {
+                stack.Pop();
+            }
+        }
+
+        Expr EncodeIntern(object obj)
+        {
+            var type = obj.GetType();
+
+            switch (Type.GetTypeCode(type))
+            {
+                case TypeCode.String:
+                    return StringV.Of((string)obj);
+
+                case TypeCode.Boolean:
+                    return BooleanV.Of((bool)obj);
+
+                case TypeCode.DateTime:
+                    return WrapDateTime((DateTime)obj);
+
+                case TypeCode.SByte:
+                case TypeCode.Int16:
+                case TypeCode.Int32:
+                case TypeCode.Int64:
+                    return LongV.Of(Convert.ToInt64(obj));
+
+                case TypeCode.Char:
+                case TypeCode.Byte:
+                case TypeCode.UInt16:
+                case TypeCode.UInt32:
+                case TypeCode.UInt64:
+                    return LongV.Of(Convert.ToInt64(obj));
+
+                case TypeCode.Single:
+                case TypeCode.Double:
+                case TypeCode.Decimal:
+                    return DoubleV.Of(Convert.ToDouble(obj));
+
+                case TypeCode.Empty:
+                    return NullV.Instance;
+
+                case TypeCode.Object:
+                    if (typeof(byte[]).IsInstanceOfType(obj))
+                        return new BytesV((byte[])obj);
+
+                    if (type.IsArray)
+                        return WrapEnumerable((IEnumerable)obj);
+
+                    if (typeof(IDictionary).IsInstanceOfType(obj))
+                        return WrapDictionary((IDictionary)obj);
+
+                    if (typeof(IEnumerable).IsInstanceOfType(obj))
+                        return WrapEnumerable((IEnumerable)obj);
+
+                    return WrapObj(obj);
+            }
+
+            return NullV.Instance;
```

Comment:
an exception would make more sense, given I tested every possible enum values in the switch